### DISCUSSION TITLE
Using a generated object as input setup({...}).createMachine

### DISCUSSION THREAD

**Discussion by justin-prather** [statelyai/xstate#5110] [Q&A] [answered]
I am trying to generate the object passed to createMachine at runtime, and i thought i could just save a base config, mutate it as needed, and then pass it to create machine. however i am getting completely swamped in type errors even doing a basic proof of concept (extracting the json object i had passed into create machine to a variable and then just passing that variable into create machine). Im really struggling with how to type the param passed into create machine, any help would be greatly appreciated

**Comment by Andarist** [comment]
Please share what you have come up with so far. It will take us less time to figure out how to help you with some initial code at hand.

**Comment by justin-prather** [comment]
```
const machineConfig = {
  context: {},
  id: "Logic Controller",
  initial: "Idle",
  states: {
    Idle: {
      on: {
        onFreshData: {
          target: "Pull Hard Inputs",
        },
        onUserData: {
          target: "Pull User Inputs",
        },
      },
    },
    "Pull Hard Inputs": {
      invoke: {
        src: "hardInputs",
        onDone: {
          target: "Compute Synthetic Inputs",
        },
      },
    },
    "Pull User Inputs": {
      invoke: {
        src: "userInputs",
        onDone: {
          target: "Compute Synthetic Inputs",
        },
      },
      description: "Pull user inputs and user overrides",
    },
    "Compute Synthetic Inputs": {
      type: "parallel",
      onDone: {
        target: "Compute Outputs",
      },
      states: {
        "New state 1": {
          initial: "runner",
          states: {
            runner: {
              invoke: {
                input: {},
                src: "actor",
                onDone: {
                  target: "finished",
                },
              },
            },
            finished: {
              type: "final",
            },
          },
        },
        "New state 2": {
          initial: "runner",
          states: {
            runner: {
              invoke: {
                input: {},
                src: "actor2",
                onDone: {
                  target: "finished",
                },
              },
            },
            finished: {
              type: "final",
            },
          },
        },
      },
    },
    "Compute Outputs": {
      type: "parallel",
      onDone: {
        target: "Push Hard Outputs",
      },
      states: {
        "New state 1": {
          initial: "runner",
          states: {
            runner: {
              invoke: {
                input: {},
                src: "actor3",
                onDone: {
                  target: "finished",
                },
              },
            },
            finished: {
              type: "final",
            },
          },
        },
        "New state 2": {
          initial: "runner",
          states: {
            runner: {
              invoke: {
                input: {},
                src: "actor4",
                onDone: {
                  target: "finished",
                },
              },
            },
            finished: {
              type: "final",
            },
          },
        },
      },
    },
    "Push Hard Outputs": {
      invoke: {
        src: "pushOutputs",
        onDone: {
          target: "Idle",
        },
      },
    },
  },
};

export const machine = setup({
  types: {
    context: {} as {},
    events: {} as Events,
  },
  actors: Actors,
}).createMachine(machineConfig);
```

**Reply by Andarist** [reply, marked as answer]
This works pretty well: [TS plaground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgZwKYwK5gDRwIIB2AngKIBuqBMA8gEYBWqAxjLoUXi9ADIQDmwJnAC+cAGZQIIOACIAHshgBDGKhkBuALAAoHUwgFFcEEqYALYAVQBhA2OB8AXHAAKSqEpDpUUZAB4YIjBUCDFjADomKFQVVABZUwsrAD4AbQAGAF04AF5EHTg4fSpUORhnBGFsArhgABNnGV4BIVsqSQAbDp8Zau1Cy2AYYCUOxoBJOu7emsVY5Aqawsnuxf7CjYM1jZ24AwAxaOQzABEVJW3dneUoPnRGlwwuuAAJdzq4cYIwDBhkGfWVyqS12BgAqmgoGdlJcroUbndyrJHs8IT5Pt9fv8+nCRDjdsDAXiQTIUR1Xu8MT8-jJYRtLGQIABrVB0nbIKBMRpmd5fanYkE7AwnAys-JEq4I+6yWzgX6oOAAZSIVDM6EEVKxANxhKB+MKuo2pKe5LRUE1NLZtQIjJZVsKHK5sgwkL5Wv1oIIIqs9o2UqRMll1IVytV6qEbppHp2hpj0bgdVQyCiwDAwy2yJNcBd6Ms-LgSgIHxz5ogFCgUHqSe1G1jMqkwaVKpgauGEcxlvFkqCYpkYHco26HRrQq9ot9-saQflcGov35I9r8bmqgWXbhMgAcqgAO4oZSqOAARlp69xg2Go0aUAwBCsUEXVxXSd9Oxvd58r92DOZYoQgtxekOwqOtAI2R1GlMGBoEfMC9jHH0zzg653ERRp7AIYBjlQOpYLA0DAIIvUAKuDCsLVBokLAwJgnQyxRjwgl4zjEjiQlI1tz3Z84AAJlPf92J2C8RjGWR33vRiHQPF8qLhcTP1k88bV-L8rjzX4QOYwCINkKDoD4rTcWFcdFLgydZDI7DcMMuEiNsmyDQc8RBis1S-R7OiCAYpy7Mc1iCLrOtAwbGc5xgBcrRo3t+w8LpUGHeNjMQgTcXM41jgpKAPjChd4wI581xSjdOP3WJj341iBkwy9RJkeSHxsgq3PqtyqttP9KrhdSkUqJz2U5SCuCgABmSS4SSjrBLM1DpRkSyKLG4ipqW5DfI2eacNauAos87zOr85a2J1LSt13UrDz4r9hKvMTbwkxrpMK-a4Ba0y1OUu03q64DEDWp8Bt0oaABZFs9b1JuQyUZoDDbrOeljDoR1anNhrados+iEueta7ICj0gseDK3iy2d5y1K0f0+or-qdPsXTMHL3VYiavzSlY1C0vGQUNQlhC0XRtFKSBYCKAwjGkPI0EwMAAAoiqip6iWKVQyhAgtkF+j1UAoKhCtEJQNfYchKBoBhmFYGpDT03w1YNuAACVmGgOo-EUSsCD4NhiE4aCoGaQRkj6YQAEp+Z0EBImiWIEnMSxUBlkxY6sNp7D4UOdCAA)

**Reply by justin-prather** [reply]
Thank you so much. I got so lost in complicated answers following the ts error messages i didnt consider something so simple. appreciate it @Andarist
